### DISCUSSION TITLE
potential browserstack-service issue or incorrect implementation on my part with wdio v8 and v9

### DISCUSSION THREAD

**Discussion by troyyer** [webdriverio/webdriverio#13607] [Q&A] [answered]
Hi,

I just have a weird behaviour I am facing that I want some smarter eyes to help identity what I am doing wrong.
I am upgrading my automation tests from V7 to v8 (and then to V9).
I run our tests on both my Local Machine as well as on the Browserstack cloud infrastructure.
I have created a dummy project with version 8 https://github.com/troyyer/wdiov8forupgradepurpose
So as part of V8 I noted that putting in the wdio.conf.js file the services: ['chromedriver'] is deprecated from the config (as part of npm package   "wdio-chromedriver-service": "8")

so the weird behaviour I am seeing (which is run with `npm run wdiowBSuser`), if I have commented out the browserstack service _but_ at the same time I leave a reference to the username and password for the browserstack account within the file, when I execute a test that I expect to run on my local machine, that test now actually runs on the browserstack cloud environment, which to me is weird as I have removed the service from the config (or I think I have). Why is this behaving this way? have I misconfigured my config file? 
```
    user: process.env.BROWSERSTACK_USERNAME || 'BROWSERSTACK_USERNAME',
    key: process.env.BROWSERSTACK_ACCESS_KEY || 'BROWSERSTACK_ACCESS_KEY',
...
...
    services: [
    //     ["browserstack", { 
    //     testObservability: false,
    //     testObservabilityOptions: {
    //         'projectName': BrowserstackProjectName,
    //         buildName: `${BrowserstackBuildName}_${process.env.ENV}`,
    //         'buildTag':[`${process.env.ENV}`,`${new Date().toLocaleDateString("en-AU")}`]
    //     },
    //   }]
    ],
```

the other weird thing is if I enable the browserstack service in config file _but_ I want the runner to still be running locally on my machine  `npm run wdiowBSservice` then this still runs on the browserstack cloud infrastructure. 
one of the features from browserstack is integrating their observability component with a test running locally on your laptop/CI system https://www.browserstack.com/docs/test-observability/quick-start/webdriverio#Tests_running_locally_or_elsewhere    
 so I wanted to have the details present in my local chrome conf file.
If someone can please correct my implementation or if it is an actual issue with the @wdio/browserstack-service I can raise it accordingly.

**Comment by christian-bromann** [comment, marked as answer]
> So as part of V8 I noted that putting in the wdio.conf.js file the services: ['chromedriver'] is deprecated from the config (as part of npm package "wdio-chromedriver-service": "8")

Yes, WebdriverIO now sets up driver for you, you can remove these driver services.

> _but_ at the same time I leave a reference to the username and password for the browserstack account within the file

If you define `user` and `key` in the WDIO config, WebdriverIO will automatically connect to the cloud provider that matches the key structure, in this case Browserstack. If you don't want that, please remove these properties from your config.

> which to me is weird as I have removed the service from the config

The service "just" enhances the integration between WebdriverIO and BrowserStack. You don't need the service to run your tests on BrowserStack.

> one of the features from browserstack is integrating their observability component with a test running locally on your laptop/CI system

I am not entirely sure how the product works but I assume you will have to provide the access key or API key to the service differently.

I hope this helps, it is better to maybe directly reach out to BS support.

**Comment by troyyer** [comment]
hmmmm in this case that implementation from version 8 doesn't work in this instance , cause even thou I have specified the **Browserstack service** along with **username and key**, in this case the test execution should still run on the **local machine**.

following their steps mentioned in their docs 
https://www.browserstack.com/docs/test-observability/quick-start/webdriverio#Tests_running_locally_or_elsewhere
```
  exports.config = {
    //...
    services: [
        ['browserstack', {
            testObservability: true,
            testObservabilityOptions: {
                user: 'myusername' || process.env.BROWSERSTACK_USERNAME,
                key: 'myaccesskey' || process.env.BROWSERSTACK_ACCESS_KEY,
                projectName: 'Specify your project name here. It should not be dynamic',
                buildName: 'Specify your job name here. For e.g. `Nightly regression`. It should not be dynamic',
                buildTag: 'Any build tag goes here. For e.g. ["Tag1","Tag2"]'
            },
        }]
    ],
    //...
  };

```
i created another npm script and ` npm run wdiowBSobs` and when run i get an error from the webdriverio detectBackend function

```
[0-0] RUNNING in chrome - file:///C:/code/wdiov8base/test/specs/test.e2e.js
[0-0] 2024-09-17T10:18:56.801Z ERROR @wdio/runner: Error: A "user" or "key" was provided but could not be connected to a known cloud service (Sauce Labs, Browerstack, Testingbot or Lambdatest). Please check if given user and key properties are correct!
[0-0]     at detectBackend (file:///C:/code/wdiov8base/node_modules/webdriverio/build/utils/detectBackend.js:96:15)
[0-0]     at getProtocolDriver (file:///C:/code/wdiov8base/node_modules/webdriverio/build/utils/driver.js:33:36)
[0-0]     at remote (file:///C:/code/wdiov8base/node_modules/webdriverio/build/index.js:44:39)       
[0-0] FAILED in chrome - file:///C:/code/wdiov8base/test/specs/test.e2e.js
2024-09-17T10:18:57.265Z INFO @wdio/cli:launcher: Run onWorkerEnd hook
2024-09-17T10:18:57.267Z INFO @wdio/cli:launcher: Run onComplete hook
```
![image](https://github.com/user-attachments/assets/ac91b4a6-daba-4f5f-b679-549d538635a0)
*note: I edited the image as I saw I had placed the user and key in the wrong section yesterday. On update to the correct way they mentioned within testObservabilityOptions, it still didnt work :'( .
please note that if i move the user and key outside of the services section the connectivity to browserstack starts to works.

I'll raise it with the browserstack team but since the error is in the  webdriverio utils file it might be a uphill discussion with them
Unfortunate for me that the latest versions of webdriverio doesnt work in this instance with Browserstack's observability feature for local execution..i'll still raise it as their docs say it fine to also put in in the services section but seems to fail... looks like i will be stuck on V8 as that still allows for **wdio-chromedriver-service** as some of our tests can only run on our internal server..

**Reply by troyyer** [reply]
I take back the previous comment I made. I _felt I_ did make the change I mentioned above and it failed initially but on a second run it was working, so I must not have saved the file when I first ran it.
So yes I found out my incorrect implementation 
Much appreciated with your feedback yesterday @christian-bromann 

This is the final config to get it working with a local execution and also push the results to browserstack's observability feature

```
const BrowserstackProjectName = 'SampleTestBuild';
const BrowserstackBuildName = 'SampleTestBuild';

exports.config = {

    runner: 'local',
    specs: [
        './test/specs/**/*.js'
    ],
    capabilities: [{
        browserName: 'chrome',
        "wdio:chromedriverOptions": {
            binary: ".\\node_modules\\chromedriver\\lib\\chromedriver\\chromedriver.exe"
        },
        acceptInsecureCerts: true,
        'goog:chromeOptions': {
            args: ["--disable-notifications","--start-maximized"]
        },
    }],

    services: [
        ["browserstack", { 
        testObservability: true,
        testObservabilityOptions: {
            user: process.env.BROWSERSTACK_USERNAME || 'BROWSERSTACK_USERNAME',
            key: process.env.BROWSERSTACK_ACCESS_KEY || 'BROWSERSTACK_ACCESS_KEY',    
            'projectName': BrowserstackProjectName,
            buildName: `${BrowserstackBuildName}_${process.env.ENV}`,
            'buildTag':[`${process.env.ENV}`,`${new Date().toLocaleDateString("en-AU")}`]
        },
      }]
    ],
    framework: 'mocha',
    reporters: ['spec'],
    mochaOpts: {
        ui: 'bdd',
        timeout: 60000
    },

  
}
```
